File: src/main/java/com/gitlab/daring/image/command/parameter/CommandParamPanel.java
package com.gitlab.daring.image.command.parameter;

import com.gitlab.daring.image.event.VoidEvent;
import net.miginfocom.swing.MigLayout;

import javax.swing.*;
import java.util.List;
import java.util.function.Consumer;

import static java.util.Collections.emptyList;

public class CommandParamPanel extends JPanel {

    public final VoidEvent applyEvent = new VoidEvent();
    List<CommandParam<?>> params = emptyList();

    public CommandParamPanel() {
        setLayout(new MigLayout("fill, wrap 2", "[right][grow,fill]", "[center]"));
    }

    public void setParams(List<CommandParam<?>> ps) {
        if (params.equals(ps)) return;
        params = ps;
        applyEvent.removeListeners();
        removeAll();
        ps.forEach(this::addParam);
        revalidate();
    }

    public void addParamChangeListener(Consumer<Void> l) {
        params.forEach(p -> p.changeEvent.addListener(l));
    }

    void addParam(CommandParam<?> p) {
        if (p.getName().isEmpty()) return;
        if (p instanceof ParamGroup) {
            addComponent(p.getName(), new JSeparator(), "");
        } else if (p instanceof NumberParam<?>) {
            addNumberParam((NumberParam<?>) p);
        } else if (p instanceof EnumParam<?>) {
            addEnumParam((EnumParam<?>) p);
        } else if (p instanceof BooleanParam) {
            addBooleanParam((BooleanParam) p);
        } else if (p instanceof FileParam) {
            addFileParam((FileParam) p);
        } else if (p instanceof StringParam) {
            addStringParam((StringParam) p);
        }
    }

    void addNumberParam(NumberParam<?> p) {
        JSlider sl = new JSlider(p.minValue.intValue(), p.maxValue.intValue(), p.getValue().intValue());
        int range = sl.getMaximum() - sl.getMinimum();
        sl.setMajorTickSpacing(range > 15 ? range / 10 : 1);
        sl.setPaintLabels(true);
        sl.addChangeListener(e -> p.setNumValue(sl.getValue()));
        JTextField f = newValueField(p, false);
        f.setColumns(5);
        addComponent(p.getName(), f, "split 2, growx 0");
        add(sl);
    }

    <T extends Enum<T>> void addEnumParam(EnumParam<T> p) {
        T[] vs = p.getJavaEnumClass().getEnumConstants();
        JComboBox<T> cb = new JComboBox<T>(vs);
        cb.setSelectedItem(p.getValue());
        cb.addActionListener(e -> p.setValue(vs[cb.getSelectedIndex()]));
        addComponent(p.getName(), cb, "");
    }

    void addBooleanParam(BooleanParam p) {
        JCheckBox b = new JCheckBox("", p.getValue());
        b.addItemListener(e -> p.setValue(b.isSelected()));
        addComponent(p.getName(), b, "");
    }

    void addFileParam(FileParam p) {
        JTextField f = newValueField(p, true);
        ParamFileChooser fc = new ParamFileChooser(p, f);
        addComponent(p.getName(), f, "split 2");
        add(fc.getOpenButton(), "growx 0");
    }

    void addStringParam(StringParam p) {
        JTextField f = newValueField(p, true);
        addComponent(p.getName(), f, "");
    }

    <T> JTextField newValueField(CommandParam<T> p, boolean editable) {
        JTextField f = new JTextField("" + p.getValue());
        if (editable) {
            f.addActionListener(e -> p.setStringValue(f.getText()));
            applyEvent.onFire(() -> p.vr = p.parseValue(f.getText()));
        } else {
            f.setEditable(false);
            p.changeEvent.onFire(() -> f.setText("" + p.getValue()));
        }
        return f;
    }

    void addComponent(String label, JComponent comp, String spec) {
        add(new JLabel(label));
        add(comp, spec);
    }

}

File: src/main/kotlin/com/gitlab/daring/image/command/parameter/CommandParamPanel.kt
package com.gitlab.daring.image.command.parameter

import com.gitlab.daring.image.event.VoidEvent
import net.miginfocom.swing.MigLayout
import java.util.Collections.emptyList
import javax.swing.*

class CommandParamPanel : JPanel() {

    @JvmField
    val applyEvent = VoidEvent()

    var params = emptyList<CommandParam<*>>()
        set(ps) {
            if (params == ps) return
            field = ps
            applyEvent.removeListeners()
            removeAll()
            ps.forEach { this.addParam(it) }
            revalidate()
        }

    init {
        layout = MigLayout("fill, wrap 2", "[right][grow,fill]", "[center]")
    }


    fun addParamChangeListener(l: (Void?) -> Unit) {
        params.forEach { it.changeEvent.addListener(l) }
    }

    fun addParam(p: CommandParam<*>) {
        if (p.name.isEmpty()) return
        when (p) {
            is ParamGroup -> addComponent(p.name, JSeparator(), "")
            is NumberParam<*> -> addNumberParam(p)
            is EnumParam<*> -> addEnumParam(p)
            is BooleanParam -> addBooleanParam(p)
            is FileParam -> addFileParam(p)
            is StringParam -> addStringParam(p)
        }
    }

    fun addNumberParam(p: NumberParam<*>) {
        val sl = JSlider(p.minValue.toInt(), p.maxValue.toInt(), p.value.toInt())
        val range = sl.maximum - sl.minimum
        sl.majorTickSpacing = if (range > 15) range / 10 else 1
        sl.paintLabels = true
        sl.addChangeListener { p.setNumValue(sl.value) }
        val f = newValueField(p, false)
        f.columns = 5
        addComponent(p.name, f, "split 2, growx 0")
        add(sl)
    }

    fun <T : EnumParam<*>> addEnumParam(p: T) {
        val vs = p.enumValues.map { "$it" }.toTypedArray()
        val cb = JComboBox(vs)
        cb.selectedItem = p.stringValue
        cb.addActionListener { p.stringValue = vs[cb.selectedIndex] }
        addComponent(p.name, cb, "")
    }

    fun addBooleanParam(p: BooleanParam) {
        val b = JCheckBox("", p.value)
        b.addItemListener { p.value = b.isSelected }
        addComponent(p.name, b, "")
    }

    fun addFileParam(p: FileParam) {
        val f = newValueField(p, true)
        val fc = ParamFileChooser(p, f)
        addComponent(p.name, f, "split 2")
        add(fc.openButton, "growx 0")
    }

    fun addStringParam(p: StringParam) {
        val f = newValueField(p, true)
        addComponent(p.name, f, "")
    }

    fun <T: Any> newValueField(p: CommandParam<T>, editable: Boolean): JTextField {
        val f = JTextField("" + p.value)
        if (editable) {
            f.addActionListener { p.stringValue = f.text }
            applyEvent.onFire { p.vr = p.parseValue(f.text) }
        } else {
            f.isEditable = false
            p.changeEvent.onFire { f.text = "" + p.value }
        }
        return f
    }

    fun addComponent(label: String, comp: JComponent, spec: String) {
        add(JLabel(label))
        add(comp, spec)
    }

}